### ISSUE TITLE
Hot reload doesn't work when @observer is used

### ISSUE THREAD

**Issue by eyousefifar** [mobxjs/mobx-react#500]
package version:
 "mobx": "^4.3.1",
 "mobx-react": "^5.2.2",
 "react": "16.3.1",
 "react-native": "0.55.4"
platform: Android
Hi, when I add @ observer to my react class, hot reloading doesn't work. it reloads the code but UI doesn't rerender. it works fine without an observer
I can provide more information if needed

**Comment by urugator** [comment]
https://mobx.js.org/best/stateless-HMR.html
https://mobx.js.org/best/decorators.html#disclamer-limitations-of-decorator-syntax
https://github.com/mobxjs/mobx/issues?utf8=%E2%9C%93&q=is%3Aissue+HMR
https://github.com/mobxjs/mobx-react/issues?utf8=%E2%9C%93&q=is%3Aissue+HMR

**Comment by eyousefifar** [comment]
@urugator tnx, but I used class components and removed decorators, and used observer like this : 
export default observer(App) where App is my components class, but as soon as I put observer to work, HMR stops working.

**Comment by luco** [comment]
+1 on this.

**Comment by theKashey** [comment]
@briedel1 - you just spotted one of RHL's limitations. I could understand why it will not work for SFC wrapper by observable (it will not, due to `componentClass` is trapped in function scope), but I don't have explanation why it does not work for classes.

Let me play with it for a while.

**Comment by theKashey** [comment]
So - there are always just 2 issues:
 - something don't get updated. HMR\RHL could not update some things. Re-rendering will not "fix" it.
 - something don't get re-rendered. RHL tries to force-update everything, but that doesn't always work. So - may be HMR/RHL did their job here, just MobX not seeing reason to update component, or visa versa. If it will be re-rendered by the App's property change - that's the case.

Anyway - don't touch anything :), I need a broken example to fix.

**Comment by briedel1** [comment]
Thanks for taking a look.

**Comment by theKashey** [comment]
So everything is badly broken. 
You may enable RHL debug mode using `setConfig`, and then see the problem
>bundle.js:918 React-hot-loader: reconciliation failed due to error TypeError: Cannot read property 'obsObj' of undefined

`this.state` - is really undefined, due to `Symbol('state value holder')` created in `makeObservableProp` is unique.

I've tried to "concrete" them, so allocate one symbol per properly name, and it fixes the problem.

@mweststrate - is there any sense to use unique symbols to shadow "well-known" props like state and props?

**Comment by theKashey** [comment]
What about just caching them? It `createSymbol` could return the same value for the same keys?
If ok - I'll double check from my side and draft a PR.

**Comment by mweststrate** [comment]
That would be the neater solution indeed :)

Op do 26 jul. 2018 om 15:51 schreef Anton Korzunov <notifications@github.com

**Comment by briedel1** [comment]
Thanks for the help and quick turnaround @theKashey !

**Comment by mweststrate** [comment]
Released as 5.2.5. Thanks for investigating and fixing @theKashey !

**Comment by reza7rm** [comment]
> Released as 5.2.5. Thanks for investigating and fixing @theKashey !
Hi all
I still have the same problem with the following versions:
    "mobx": "^4.5.0",
    "mobx-react": "^5.3.3",
    "mobx-state-tree": "^3.7.0",

**Comment by Kruemelkatze** [comment]
> Released as 5.2.5. Thanks for investigating and fixing @theKashey !
> Hi
> I still have the same problem with the following versions:
> "mobx": "^4.5.0",
> "mobx-react": "^5.3.3",
> "mobx-state-tree": "^3.7.0",

@reza7rm 
I had the same issue, but downgrading mobx-react to ```5.2.8``` seems to work. <5.2.5 and >5.2.8 wouldn't work. 

Our Dependencies:
```js
"mobx": "4.5.2", // Thanks to IE
"mobx-react": "5.2.8",
"react": "^16.6.0",
"react-dom": "^16.6.0",
"react-hot-loader": "~4.3.12",
```

With this setup, I got a nested component (with state, an injected store and @ observer) to update both on store changes (via mobx) and file changes like changing text (via HMR). Our test component:
```jsx
import React, { Component } from 'react'
import { observer, inject } from 'mobx-react';

@inject("permissionStore")
@observer
class Dummy extends Component {

    state = {
        switchedIcon: false,
    }

    render() {
        let userName = `${this.props.permissionStore.user.givenName} ${this.props.permissionStore.user.surName}`;
        return (
            <div>
                <h2>{userName}</h2>

                <button onClick={() => this.setState({ switchedIcon: !this.state.switchedIcon })}>Switch Icon</button>
                <FontAwesomeIcon icon={this.state.switchedIcon ? 'minus' : 'plus'} />
            </div>
        )
    }
}

export default Dummy
```

This setup did not work with <5.2.5 and none of the ~5.3.x releases of mobx-react. But as I said, 5.2.8 works fine.

**Comment by zfalen** [comment]
Strange thing I noticed over here, experiencing the exact same behavior. MobX observer tag is preventing **child** components from hot reloading. 

_However_ - I had the same setup _(except using webpack-hot-middleware)_ working on a totally different project, and started looking for other differences in config.... 

It turns out that **removing the `import "babel-polyfill";` line totally fixed the issue.** I was using that import to enable async / await but turning it off completely re-enabled HMR. Not sure what the implications are, if any, but maybe a useful clue?
@theKashey @mweststrate 

_Note:_ HMR _and_ observer are confirmed working at the root and child level, as expected. Local state is preserved during HMR reload. I did _not_ need to downgrade libraries to fix.

**Dependencies:**
`"mobx": "4.5.2",
"mobx-react": "5.2.8",
"react": "^16.6.0",
"react-dom": "^16.6.0",
"react-hot-loader": "~4.3.12",`

**Top-level react component _(with react-hot-loader setup)_:** 
`class App extends Component { ... } export default hot(module)(App);`

**Child react component:** 
`
@observer
export default class SomeComponent extends Component { ... }`

**Comment by theKashey** [comment]
@zfalen `babel-polyfill` should be the first import in a whole project. Was it?
Having it imported not in the first place could lead to undefined behaviour.

Easiest example
```js
import 'react'
import 'babel-polyfill'
import 'react-dom';
🔥 react and react-dom got different "Symbol" in IE11, everything is broken
```

PS: please ensure that you have `modules:false` in your babel config, or react-hot-loader will import itself(and react) in a first place, probably breaking some stuff.

**Comment by zfalen** [comment]
@theKashey **TLDR I found the issue - _dont ever_ use an arrow function on lifecyle methods** _(or any other format for that matter. Even if there was a fire)_ 

Here are some longwinded tweaking results that may or may not help you or other people experiencing the issue :)  

`babel-polyfill` is the first line in the index file: 

```
import "babel-polyfill";
import "../sass/index.scss";
import React from "react";
import ReactDOM from "react-dom";

// hot(module)(App) is in this file 
import App from "./components/app";

ReactDOM.render(<App />, document.getElementById("app"));
```

I also tried adding `modules: false` directly in the webpack config _(though it didnt seem to make any difference in behavior):_
  ```
    {
        test: /\.(jsx)$/,
        exclude: /(node_modules|bower_components)/,
        use: [
          {
            loader: "babel-loader?cacheDirectory",
            options: {
              cacheDirectory: true,
              presets: [
                ["@babel/preset-env", { modules: false }],
                "@babel/react",
                "mobx"
              ],
              plugins: ["react-hot-loader/babel"]
            }
          }
        ]
      },
``` 
However, I either totally misreported what I was experiencing at first, or I am on drugs. Testing this morning things started breaking again, which led me down another debug path. As it turns out, **`babel-polyfill` didn't affect HMR today.** Instead, what I noticed was that I had a lifecycle method declaration in the child component which was created as an _arrow_ function for some unknown reason I cannot explain. Eg:

  ```
componentDidMount = () => {
    fetch("http://0.0.0.0:3333/")
      .then(res => res.json())
      .then(json => (this.serverData = json));
  }
```

This did not work. Commenting out that line while trying to get things to work again _did_ get HMR working. Voilé! A few more tweaks / experiments later, and I notice:

  `componentDidMount = function() { ... }`
Did not work either. 

  `componentDidMount() { ... }`
_**Does work**_ as expected! As one should hope, since this is the right way to do that.

Some more variations and their results: 

  ```
@action
  componentDidMount() {
    fetch("http://0.0.0.0:3333/")
      .then(res => res.json())
      .then(json => (this.serverData = json));
  }
```
**Works.**


 ```
 async componentDidMount() {
    const res = await fetch("http://0.0.0.0:3333/");
    const json = await res.json();
    this.serverData = json;
  }
```
**Works.** _(w/polyfill)_

 ```
 @action updateServerData(val) {
    this.serverData = val;
  }

  async componentDidMount() {
    ...
    this.updateServerData(json);
  }
```
**Works.** _(w/polyfill)_


Awesome. Seem to have found the issue. Doing weird and unnecessary things to standard React lifecycle methods prevent the component from hot reloading. It was _**not**_ an issue with MobX `observer`. Notably, it happens with `render` too:
  ```
render = () => (
    <div>
      <h4>I Am A SubComponent</h4>
      ...
    </div>
  );
```
**Does Not Work.**

Although stateless components written with an arrow function _**do**_ work as expected with HMR:
```
const Stateless = ({ someProp }) => <h2>{someProp}</h2>;
...
<Stateless someProp={'hello'} />
```

**Comment by theKashey** [comment]
We had an issue with _render-as-arrow-function_, but, yeah, we never tested _componentDidMount_-as-arrow-function 🙀

**Comment by wuifdesign** [comment]
I'm still having problems with observer decorator in React Native.

```
"mobx": "^5.9.0",
"mobx-react": "^5.4.3",
"react": "16.6.3",
"react-native": "0.57.8"
```

React Native HMR not working when using a decorator like this: 

```
import {observer} from 'mobx-react/native'

@observer
export default class App extends Component<Props> {
  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.welcome}>Demo Text</Text>
      </View>
    );
  }
}
```

but this works when calling as a funtion:

```
import {observer} from 'mobx-react/native'

class App extends Component<Props> {
  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.welcome}>Demo Text</Text>
      </View>
    );
  }
}

export default observer(App);
```

**Comment by theKashey** [comment]
@wuifdesign - react-native uses quite old version of RHL, and there is no way for me to help you here.

**Comment by ConneXNL** [comment]
@theKashey @mweststrate @eyousefifar Can anyone re-open this issue?
You still need to explicitly use  "mobx-react": "5.2.8" in order for the issue to not occur.

Apparently the fix implemented in 5.2.5 broke again in 5.3 >.
